README.md

```diff
@@ -175,18 +175,26 @@ The result of executing a GraphQL operation may contain partial data as well as
 Errors that happen during execution of the GraphQL operation typically become part of the result,
 as long as the server is still able to produce a well-formed response.
 
-In case of errors that completely prevent the successful execution of the request, the server SHOULD respond
-with the appropriate status code depending on the concrete error condition. For example:
+In case of errors that completely prevent the successful execution of the request, the server MUST respond
+with the appropriate status code depending on the concrete error condition.
 
-- `400` (Bad Request), in case of a validation error
-- `401` (Unauthorized), if an unauthenticated client is not allowed to access the schema
-- `500` (Internal Server Error), for unexpected failures in the server
+There are four types of validation errors:
 
-As an example, the way authorization errors may be handled illustrates the different error conditions.
-If a client is forbidden from making any GraphQL requests at all, the server may respond with
-status code `403` (Forbidden) and not return any GraphQL data.
-In contrast to that, a client might be allowed to access the schema, but requests some restricted fields.
-Given the server can provide at least partial `data`, the response would have status code `200` (OK) and 
-include `errors` as part of the GraphQL response.
+1. Unparseable or invalid request body, for example `NONSENSE` or `{"qeury": "{__typename}"}`
+2. GraphQL Specification document validation errors (valid syntax, no loops in fragments, etc)
+3. Custom document validations that can be performed before execution (e.g. graphql-depth-limit)
+4. Runtime validations performed by resolvers (i.e. during the execution of the GraphQL operation).
 
-It is RECOMMENDED to use the same error codes as the [reference implementation](https://github.com/graphql/express-graphql).
+The error types 1. and 2. prevent execution of the request entirely and MUST result in
+status code `400` (Bad Request). Types 3. and 4. may still allow partial execution and
+MUST result in status code `200`.
+
+If an unauthenticated client is not allowed to access the schema, the server MUST respond with
+status code `401` (Unauthorized).
```

Comment:
At the moment this precludes the use of 403 for unauthenticated clients. Further, one of 401, 403 or 404 should be permitted in these situations, they each have different security implications.

```suggestion
If an unauthenticated client is not allowed to access the schema, but an authenticated client might be permitted, the server SHOULD respond with
status code `401` (Unauthorized).
```

Reply:
> At the moment this precludes the use of 403 for unauthenticated clients.

Missing or failed authentication is precisely what 401 is meant for, or is it not? 403 and 404 each mean something different.

Concerning the use of `SHOULD` or `MUST`: How about we use `MUST` for things that are important for interoperability with clients? In this case, it seems like clients might be able to do something useful with a 401, e.g. redirecting to a login page.

Reply:
I should expand on the security implications I mentioned. Statuses 401 and 403 (generally) confirm the resource exists and that you're not allowed to see it, 404 does not. So some servers choose to use 404 to get bots/scripts to "stop looking" - the page doesn't seem to exist so there's no point trying to hack it. Have a read of: https://www.bennadel.com/blog/2400-handling-forbidden-restful-requests-401-vs-403-vs-404.htm

It should be up to the developer to decide which of these status codes is relevant to their API design.